lib/src/git.rs

```diff
@@ -1330,65 +1335,191 @@ fn push_refs(
+        callbacks.push_update_reference(|refname, status| {
+            // The status is Some if the ref update was rejected
+            if status.is_none() {
+                remaining_remote_refs.remove(refname);
+            }
+            Ok(())
+        });
+        push_options.remote_callbacks(callbacks);
+        remote
+            .push(refspecs, Some(&mut push_options))
+            .map_err(|err| match (err.class(), err.code()) {
+                (git2::ErrorClass::Reference, git2::ErrorCode::NotFastForward) => {
+                    GitPushError::NotFastForward
+                }
+                _ => GitPushError::InternalGitError(err),
+            })
+    };
+    if !failed_push_negotiations.is_empty() {
+        // If the push negotiation returned an error, `remote.push` would not
+        // have pushed anything and would have returned an error, as expected.
+        // However, the error it returns is not necessarily the error we'd
+        // expect. It also depends on the exact versions of `libgit2` and
+        // `git2.rs`. So, we cannot rely on it containing any useful
+        // information. See https://github.com/rust-lang/git2-rs/issues/1042.
+        assert!(push_result.is_err());
+        failed_push_negotiations.sort();
+        Err(GitPushError::RefInUnexpectedLocation(
+            failed_push_negotiations,
         ))
+    } else {
+        push_result?;
+        if remaining_remote_refs.is_empty() {
+            Ok(())
+        } else {
+            Err(GitPushError::RefUpdateRejected(
+                remaining_remote_refs
+                    .iter()
+                    .sorted()
+                    .map(|name| name.to_string())
+                    .collect(),
+            ))
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum PushAllowReason {
+    NormalMatch,
+    ExceptionalFastforward,
+    UnexpectedNoop,
+}
+
+fn allow_push(
+    index: &dyn Index,
+    actual_remote_location: Option<&CommitId>,
+    expected_remote_location: Option<&CommitId>,
+    local_location: Option<&CommitId>,
+) -> Result<PushAllowReason, ()> {
+    if actual_remote_location == expected_remote_location {
+        return Ok(PushAllowReason::NormalMatch);
+    }
+
+    // If the remote ref is in an unexpected location, we still allow some
+    // pushes, based on whether `jj git fetch` would result in a conflicted
+    // ref.
+    //
+    // For `merge_ref_targets` to work, `actual_remote_location` must be a commit
+    // that we locally know about. This does not lose any generality since for
+    // `merge_ref_targets` to resolve to `local_target`, it is
+    // necessary (but not sufficient) for the push to be a fast-forward or a noop:
+    // local_target is either a descendant of actual_remote_location or equal to it.
+    // TODO(ilyagr): Use if `Some(id) = ... && ...` when that stabilizes.
+    if actual_remote_location
+        .map(|id| !index.has_id(id))
+        .unwrap_or(false)
+    {
```

Comment:
nit: you can use `.map_or()` to make it one line.

btw, I would write it as `!action_remote_location.map_or(true, ..)` to clarify that `None` is considered truthy in this context.